src/ast/ast_validations.c

```diff
@@ -658,121 +666,108 @@ static AST_Validation _Validate_MATCH_Clauses(const AST *ast, char **reason) {
-	if(with_clause_count == 0) return AST_VALID;
+// Verify that MERGE doesn't redeclare bound nodes.
+static AST_Validation _ValidateMergeNode(const cypher_astnode_t *entity, rax *defined_aliases,
+										 char **reason) {
+	if(raxSize(defined_aliases) == 0) return AST_VALID;
 
-	rax *with_projections = raxNew();
-	uint clause_count = cypher_ast_query_nclauses(ast->root);
-	AST_Validation res = AST_VALID;
+	const cypher_astnode_t *identifier = cypher_ast_node_pattern_get_identifier(entity);
+	if(identifier == NULL) return AST_VALID;
 
-	// Visit all clauses in order
-	for(uint i = 0; i < clause_count; i ++) {
-		const cypher_astnode_t *clause = cypher_ast_query_get_clause(ast->root, i);
-		cypher_astnode_type_t type = cypher_astnode_type(clause);
-		if(type == CYPHER_AST_WITH) {
-			// Verify that functions invoked by the WITH clause are valid.
-			res = _ValidateFunctionCalls(clause, reason, true);
-			if(res == AST_INVALID) break;
-
-			// Collect projected aliases from WITH clauses into the same triemap
-			_AST_GetWithAliases(clause, with_projections);
-		} else if(type == CYPHER_AST_MATCH) {
-			// Verify that no path in a MATCH pattern reuses a WITH projection
-			const cypher_astnode_t *pattern = cypher_ast_match_get_pattern(clause);
-			uint path_count = cypher_ast_pattern_npaths(pattern);
-			for(uint j = 0; j < path_count; j ++) {
-				const cypher_astnode_t *path = cypher_ast_pattern_get_path(pattern, j);
-				res = _ValidateWithEntitiesOnPath(path, with_projections, reason);
-				if(res == AST_INVALID) break;
-			}
-		} else if(type == CYPHER_AST_CREATE) {
-			// Verify that no path in a CREATE pattern reuses a WITH projection
-			const cypher_astnode_t *pattern = cypher_ast_create_get_pattern(clause);
-			uint path_count = cypher_ast_pattern_npaths(pattern);
-			for(uint j = 0; j < path_count; j ++) {
-				const cypher_astnode_t *path = cypher_ast_pattern_get_path(pattern, j);
-				res = _ValidateWithEntitiesOnPath(path, with_projections, reason);
-				if(res == AST_INVALID) break;
-			}
+	const char *alias = cypher_ast_identifier_get_name(identifier);
+	if(raxFind(defined_aliases, (unsigned char *)alias, strlen(alias)) == raxNotFound) {
+		// If the entity is unaliased or not previously bound, it cannot be redeclared.
+		return AST_VALID;
+	}
 
-		} else if(type == CYPHER_AST_MERGE) {
-			// Verify that the single path in a MERGE clause doesn't reuse a WITH projection
-			const cypher_astnode_t *path = cypher_ast_merge_get_pattern_path(clause);
-			res = _ValidateWithEntitiesOnPath(path, with_projections, reason);
-			if(res == AST_INVALID) break;
-		}
+	// If the entity is already bound, the MERGE pattern should not introduce labels or properties.
+	if((cypher_ast_node_pattern_nlabels(entity) > 0) ||
+	   cypher_ast_node_pattern_get_properties(entity)) {
+		asprintf(reason, "The bound node '%s' can't be redeclared in a MERGE clause", alias);
+		return AST_INVALID;
 	}
```

Comment:
Ugh, why is that invalid? We don't cover this currently, nor do I think the validation to do so would be pretty, since this is perfectly valid:
`MATCH (a) MERGE (a)-[:e]->()`
Do you think this is an important case to cover?